### PR TITLE
Fix #7904: update proofview env after ltac constr:()

### LINKED ISSUE #4781: [abstract] typeclass resolution hints give very strange behavior (and an abundance of Anomaly: Uncaught exception Not_found. Please report.) (regression since 8.4)
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4781
From: @JasonGross
Reported version: 8.6

See also: [BZ#4496](https://github.com/coq/coq/issues?q=is%3Aissue%20%22Original%20bug%20ID%3A%20BZ%234496%22)
See also: [BZ#4493](https://github.com/coq/coq/issues?q=is%3Aissue%20%22Original%20bug%20ID%3A%20BZ%234493%22)
See also: [BZ#4863](https://github.com/coq/coq/issues?q=is%3Aissue%20%22Original%20bug%20ID%3A%20BZ%234863%22)

### LINKED ISSUE #7904: Strange error message "Error: Variable x should be bound to a term but is bound to a constr."
#### Version

8.8.0

#### Description of the problem

I was taking another look at #4781, and discovered this:
```coq

Class abstract_term {T} (x : T) := by_abstract_term : T.
Hint Extern 0 (@ abstract_term ?T ?x) => force_clear; change T; abstract (exact x) : typeclass_instances.

Goal True.
  let term := constr:(I) in
  let T := type of term in
  let x := constr:((_ : abstract_term term) : T) in
  let x := match constr:(x) with ?y => y end in
  pose x. (* Error: Variable x should be bound to a term but is bound to a constr. *)
```

### PR DESCRIPTION
**Pull request by SkySkimmer**
(in case of side effects)

I hit the same code path in SProp (there it was a Not_found anomaly) so I need this fix. Hopefully someday we will have a more general solution for #9146 bugs though.

Fixes #4781.
Fixes #7904.

### CODE REVIEW

File: plugins/ltac/tacinterp.ml [hunk 1]
```diff
@@ -1957,7 +1957,8 @@ let lifts f = (); fun ist x -> Ftactic.enter begin fun gl ->
   let sigma = Proofview.Goal.sigma gl in
   let (sigma, v) = f ist env sigma x in
   Proofview.tclTHEN (Proofview.Unsafe.tclEVARS sigma)
-  (Ftactic.return v)
+    (Proofview.tclTHEN (Proofview.Unsafe.tclSETENV (Global.env()))
+       (Ftactic.return v))
```
Comment (ppedrot) [inline plugins/ltac/tacinterp.ml]: Please add a FIXME comment here.

**Review by ppedrot — CHANGES_REQUESTED** [review]

File: plugins/ltac/tacinterp.ml (hunk 1 above)
Comment (SkySkimmer) [reply plugins/ltac/tacinterp.ml]: Is there a need for a fixme? It seems like every tclSETENV will go away if we fix side effects, and won't if we don't.

Comment (ppedrot) [reply plugins/ltac/tacinterp.ml]: `tclSETENV` is legit, the dreaded call is `tclSETENV + Global.env ()`.

Comment (SkySkimmer) [reply plugins/ltac/tacinterp.ml]: OK, comment added.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Should this be backported to 8.9?

**Pull request merged**